test/libsolidity/SolidityNameAndTypeResolution.cpp

```diff
@@ -5078,6 +5078,60 @@ BOOST_AUTO_TEST_CASE(invalid_address_length)
 	CHECK_WARNING(text, "checksum");
 }
 
+BOOST_AUTO_TEST_CASE(shadowing_builtins_with_functions)
+{
+	char const* text = R"(
+		contract C {
+			function keccak256() {}
+		}
+	)";
+	CHECK_WARNING(text, "Shadowing builtin symbol");
+}
+
+BOOST_AUTO_TEST_CASE(shadowing_builtins_with_variables)
+{
+	char const* text = R"(
+		contract C {
+			function f() {
+				uint msg;
+			}
+		}
+	)";
+	CHECK_WARNING(text, "Shadowing builtin symbol");
+}
+
+BOOST_AUTO_TEST_CASE(shadowing_builtins_with_events)
+{
+	char const* text = R"(
+		contract C {
+			event keccak256();
+		}
+	)";
+	CHECK_WARNING(text, "Shadowing builtin symbol");
+}
+
+BOOST_AUTO_TEST_CASE(shadowing_builtins_with_imports)
+{
+	char const* text = R"(
+	        import * as msg from "B.sol";
```

Comment:
Perhaps also add something like `import {msg, block} as X from "B.sol"` - that should also be fine.